spirv_glsl.cpp

```diff
@@ -725,44 +725,39 @@ void CompilerGLSL::emit_header()
 
 	if (!options.vulkan_semantics)
 	{
-		using Supp = KHR_shader_subgroupSupportHelper;
+		using Supp = ShaderSubgroupSupportHelper;
 
 		Supp::Result result = shader_subgroup_supporter.resolve();
 		for (uint32_t ft = 0u; ft < Supp::FeatureCount; ++ft)
 		{
 			Supp::Feature feature = static_cast<Supp::Feature>(ft);
-			if (!shader_subgroup_supporter.isFeatureRequested(feature))
+			if (!shader_subgroup_supporter.is_feature_requested(feature))
 				continue;
 
-			auto exts = shader_subgroup_supporter.getCandidatesForFeature(feature, result);
+			auto exts = shader_subgroup_supporter.get_candidates_for_feature(feature, result);
 			if (exts.empty())
 				continue;
 
-			statement("#ifdef GL_KHR_shader_subgroup");
-			statement("#extension GL_KHR_shader_subgroup : require");
 			for (auto it = exts.begin(); it != exts.end(); ++it)
 			{
-				std::string name = Supp::getExtensionName(*it);
-				auto extraNames = Supp::getExtraRequiredExtensionNames(*it);
-				if (extraNames.empty())
-					statement("#elif defined(", name, ")");
-				else
-				{
-					std::string extra;
-					for (const auto &e : extraNames)
-						extra += " && defined(" + e + ")";
-					statement("#elif defined(", name, ")", extra);
-				}
+				const Supp::Candidate ext = *it;
+
+				std::string name = Supp::get_extension_name(ext);
+				std::string extra = Supp::get_extra_required_extension_checks(ext);
+				auto extraNames = Supp::get_extra_required_extension_names(ext);
+				statement(it != exts.begin() ? "#elif" : "#if", " defined(", name, ")",
+				          extra.empty() ? "" : (" && " + extra));
 				for (const auto &e : extraNames)
-					statement("#extension ", e, " : required");
+					statement("#extension ", e, " : enable");
 				statement("#extension ", name, " : required");
 			}
-			if (!exts.empty()) // if getCandidatesForFeature() returns empty list, it means no extensions are needed
+			// if get_candidates_for_feature() returns empty list, it means no extensions are needed
+			if (!exts.empty() || Supp::can_feature_be_implemented_wo_extensions(feature))
```

Comment:
btw right now it looks like you require NV extension for LSB and MSB of the ballot,

Reply:
yes, because in case of NV you have to perform findLSB/MSB on just one uint instead of two. It's your idea from https://github.com/KhronosGroup/SPIRV-Cross/issues/1352

Reply:
yeah, but if NV extension is not available, we should sitll have working subgroup ballot LSB/MSB

Reply:
thats why there is `can_feature_be_implemented_wo_extensions` function ;)